program-runtime/src/invoke_context.rs

```diff
@@ -1519,4 +1519,47 @@ mod tests {
             );
         }
     }
+
+    #[test]
+    fn test_define_default_cost_for_builtins() {
+
+        // Motivation:
+        // Default compute units cost for each built-in programs could be
+        // encapsulated within program-runtime, accessible to both invoke_context
+        // (to make builtins to consume CUs) and by runtime::cost_model.
+        //
+        lazy_static::lazy_static! {
```

Comment:
This static HashMap currently exists in `runtime`, not directly accessible by `program-runtime::invoke_context`. IMO it makes sense to define native's default cost in `program-runtime`, where `compute_budget` declares costs for various actions.  Other option is to pass the default costs HashMap from `runtime` via builtin programs call chain to invoke_context.

Reply:
Native programs are defined / built on top of `program-runtime`, so it makes more sense to me to define their costs somewhere above `program-runtime`.

 * the compute budget defines costs for syscalls that are implemented in the program-runtime
 * the runtime defines costs for its native programs
 * the runtime currently passes native programs down to the invoke context

You could even have a cost field in `invoke_context::BuiltinProgram`, and deduct straight from there, and that way it'll be impossible to define a new builtin program and forget to associate a cost. What do you think?

Reply:
> so it makes more sense to me to define their _costs_  somewhere above program-runtime.

While I think IDs can be declared elsewhere, but totally agree _costs_ should be defined above program-runtime, 

I like the idea a lot for builtin program to carry a mandatory cost field. That removes the extra HashMap<program_id, cost> which is hard to maintain separately. Also have builtin developer to identify its default cost is way better than what's been done so far. (can also promote dev to implement more complex builtins to properly consume units).

I made POC PR https://github.com/solana-labs/solana/pull/30639 , is this somewhat close to what's in your mind?

Reply:
If Address-lookup-table-program is the new way to go, what's our plan for existing builtins that somewhat resides in solana-program/SDk? Will they be refactored into self-contain crate just like ALT?

Reply:
Ok, good to know. Thanks for the explanation!